<?xml version="1.0"?>
<robot name="gantry_robot" xmlns:xacro="http://www.ros.org/wiki/xacro">
  
  <!-- Base Link -->
  <link name="base_link">
    <visual>
      <geometry>
        <box size="4 4 0.1"/>
      </geometry>
      <material name="grey">
        <color rgba="0.5 0.5 0.5 1.0"/>
      </material>
    </visual>
    <collision>
      <geometry>
        <box size="4 4 0.1"/>
      </geometry>
    </collision>
    <!-- Inertial properties -->
    <inertial>
      <mass value="5.0"/>
      <origin xyz="0 0 0"/>
      <inertia ixx="8.33" ixy="0.0" ixz="0.0"
               iyy="8.33" iyz="0.0"
               izz="0.17"/>
    </inertial>
  </link>

  <!-- x_axis_link -->
  <link name="x_axis_link">
    <visual>
      <geometry>
        <box size="4 0.1 0.1"/>
      </geometry>
      <material name="red">
        <color rgba="1.0 0.0 0.0 1.0"/>
      </material>
    </visual>
    <collision>
      <geometry>
        <box size="4 0.1 0.1"/>
      </geometry>
    </collision>
    <!-- Inertial properties -->
    <inertial>
      <mass value="1.0"/>
      <origin xyz="0 0 0"/>
      <inertia ixx="0.03" ixy="0.0" ixz="0.0"
               iyy="1.33" iyz="0.0"
               izz="1.33"/>
    </inertial>
  </link>

  <!-- y_axis_link -->
  <link name="y_axis_link">
    <visual>
      <geometry>
        <box size="0.1 4 0.1"/>
      </geometry>
      <material name="green">
        <color rgba="0.0 1.0 0.0 1.0"/>
      </material>
    </visual>
    <collision>
      <geometry>
        <box size="0.1 4 0.1"/>
      </geometry>
    </collision>
    <!-- Inertial properties -->
    <inertial>
      <mass value="1.0"/>
      <origin xyz="0 0 0"/>
      <inertia ixx="1.33" ixy="0.0" ixz="0.0"
               iyy="0.03" iyz="0.0"
               izz="1.33"/>
    </inertial>
  </link>

  <!-- support_block1 -->
  <link name="support_block1">
    <visual>
      <geometry>
        <box size="0.1 0.1 2"/>
      </geometry>
      <material name="blue">
        <color rgba="0.0 0.0 1.0 1.0"/>
      </material>
    </visual>
    <collision>
      <geometry>
        <box size="0.1 0.1 2"/>
      </geometry>
    </collision>
    <!-- Inertial properties -->
    <inertial>
      <mass value="0.5"/>
      <origin xyz="0 0 0"/>
      <inertia ixx="0.02" ixy="0.0" ixz="0.0"
               iyy="0.02" iyz="0.0"
               izz="0.0"/>
    </inertial>
  </link>

  <!-- support_block2 -->
  <link name="support_block2">
    <visual>
      <geometry>
        <box size="0.1 0.1 2"/>
      </geometry>
      <material name="blue">
        <color rgba="0.0 0.0 1.0 1.0"/>
      </material>
    </visual>
    <collision>
      <geometry>
        <box size="0.1 0.1 2"/>
      </geometry>
    </collision>
    <!-- Inertial properties -->
    <inertial>
      <mass value="0.5"/>
      <origin xyz="0 0 0"/>
      <inertia ixx="0.02" ixy="0.0" ixz="0.0"
               iyy="0.02" iyz="0.0"
               izz="0.0"/>
    </inertial>
  </link>

  <!-- welding_head_base -->
  <link name="welding_head_base">
    <visual>
      <geometry>
        <cylinder radius="0.1" length="0.2"/>
      </geometry>
      <material name="silver">
        <color rgba="0.75 0.75 0.75 1.0"/>
      </material>
    </visual>
    <collision>
      <geometry>
        <cylinder radius="0.1" length="0.2"/>
      </geometry>
    </collision>
    <!-- Inertial properties -->
    <inertial>
      <mass value="0.2"/>
      <origin xyz="0 0 0"/>
      <inertia ixx="0.001" ixy="0.0" ixz="0.0"
               iyy="0.001" iyz="0.0"
               izz="0.0"/>
    </inertial>
  </link>

  <!-- Joints -->
  
  <!-- x_axis_joint -->
  <joint name="x_axis_joint" type="prismatic">
    <parent link="y_axis_link"/>
    <child link="x_axis_link"/>
    <origin xyz="-2 0 0" rpy="0 0 0"/>
    <axis xyz="0 1 0"/> 
    <limit lower="-1.2" upper="1.2" effort="100" velocity="0.5"/>
  </joint>

  <!-- y_axis_joint -->
  <joint name="y_axis_joint" type="prismatic">
    <parent link="base_link"/>
    <child link="y_axis_link"/>
    <origin xyz="2 0 1" rpy="0 0 0"/>
    <axis xyz="0 0 1"/>
    <limit lower="-0.5" upper="0.5" effort="100" velocity="0.1"/>
  </joint>

  <!-- support_joint1 -->
  <joint name="support_joint1" type="fixed">
    <parent link="base_link"/>
    <child link="support_block1"/>
    <origin xyz="2 2 1" rpy="0 0 0"/>
  </joint>

  <!-- support_joint2 -->
  <joint name="support_joint2" type="fixed">
    <parent link="base_link"/>
    <child link="support_block2"/>
    <origin xyz="2 -2 1" rpy="0 0 0"/>
  </joint>

  <!-- welding_head_joint -->
  <joint name="welding_head_joint" type="prismatic">
    <parent link="x_axis_link"/>
    <child link="welding_head_base"/>
    <origin xyz="-2 0 0" rpy="0 0 0"/>
    <axis xyz="1 0 0"/>
    <limit lower="0.5" upper="3.5" effort="10" velocity="1"/>
  </joint>

  <ros2_control name="GazeboSystem" type="system">
    <hardware>
      <plugin>gazebo_ros2_control/GazeboSystem</plugin>
    </hardware>
    <joint name="x_axis_joint">
      <command_interface name="position">
        <param name="min">{-1.2}</param>
        <param name="max">{1.2}</param>
      </command_interface>
      <state_interface name="position"/>
      <state_interface name="velocity"/>
      <state_interface name="effort"/>
    </joint>
    <joint name="y_axis_joint">
      <command_interface name="position">
        <param name="min">{-0.5}</param>
        <param name="max">{0.5}</param>
      </command_interface>
      <state_interface name="position"/>
      <state_interface name="velocity"/>
      <state_interface name="effort"/>
    </joint>
    <joint name="welding_head_joint">
      <command_interface name="position">
        <param name="min">{0.5}</param>
        <param name="max">{3.5}</param>
      </command_interface>
      <state_interface name="position"/>
      <state_interface name="velocity"/>
      <state_interface name="effort"/>
    </joint>
    
  </ros2_control>
  
  <gazebo>
    <plugin filename="libgazebo_ros2_control.so" name="gazebo_ros2_control">
      <parameters>/home/mustafa/ws_humble_urdf/install/urdf_bringup/share/urdf_bringup/config/controller_configuration.yaml</parameters>
    </plugin>
  </gazebo>

  <!-- Camera Link -->
  <link name="camera_link">
    <visual>
      <geometry>
        <box size="0.1 0.1 0.1"/>
      </geometry>
      <material name="black">
        <color rgba="0.0 0.0 0.0 1.0"/>
      </material>
    </visual>
    <collision>
      <geometry>
        <box size="0.1 0.1 0.1"/>
      </geometry>
    </collision>
    <!-- Inertial properties -->
    <inertial>
      <mass value="0.1"/>
      <origin xyz="0 0 0"/>
      <inertia ixx="0.0001" ixy="0.0" ixz="0.0"
               iyy="0.0001" iyz="0.0"
               izz="0.0001"/>
    </inertial>
  </link>

  <!-- Camera Joint -->
  <joint name="camera_joint" type="fixed">
    <parent link="base_link"/>
    <child link="camera_link"/>
    <origin xyz="-3 0 1" rpy="0 0 0"/>
  </joint>

  <!-- Camera Sensor -->
  <gazebo reference="camera_link">
    <sensor type="camera" name="camera_sensor">
      <pose> 0 0 0 0 0 0 </pose>
      <visualize>true</visualize>
      <update_rate>10</update_rate>
      <camera>
        <horizontal_fov>1.089</horizontal_fov>
        <image>
          <width>640</width>
          <height>480</height>
          <format>R8G8B8</format>
        </image>
        <clip>
          <near>0.1</near>
          <far>100.0</far>
        </clip>
      </camera>
      <plugin name="camera_controller" filename="libgazebo_ros_camera.so">
        <frame_name>camera_link</frame_name>
      </plugin>
    </sensor>
  </gazebo>


      
</robot> 
